packages/db/postgres/functions.sql

```diff
@@ -120,48 +125,39 @@ CREATE OR REPLACE FUNCTION create_upload(data json) RETURNS TEXT
 AS
 $$
 DECLARE
-  backup_url TEXT;
   inserted_upload_id BIGINT;
+
 BEGIN
   -- Set timeout as imposed by heroku
   SET LOCAL statement_timeout = '30s';
 
   PERFORM create_content(data);
 
-  insert into upload (user_id,
+  insert into upload as upld (user_id,
                       auth_key_id,
                       content_cid,
                       source_cid,
                       type,
                       name,
                       inserted_at,
-                      updated_at)
+                      updated_at,
+                      backup_urls)
   values ((data ->> 'user_id')::BIGINT,
             (data ->> 'auth_key_id')::BIGINT,
             data ->> 'content_cid',
             data ->> 'source_cid',
             (data ->> 'type')::upload_type,
             data ->> 'name',
             (data ->> 'inserted_at')::timestamptz,
-            (data ->> 'updated_at')::timestamptz)
+            (data ->> 'updated_at')::timestamptz,
+            json_arr_to_jsonb_element_array(data -> 'backup_urls'))
   ON CONFLICT ( user_id, source_cid ) DO UPDATE
     SET "updated_at" = (data ->> 'updated_at')::timestamptz,
         "name" = data ->> 'name',
-        "deleted_at" = null
+        "deleted_at" = null,
+        "backup_urls" = array_cat(upld.backup_urls, json_arr_to_jsonb_element_array(data -> 'backup_urls'))
```

Comment:
I haven't had a whole lot of time to dig into this so please take what I'm going to say with a pinch of salt.

With this implementation, if a user uploads the same CID multiple times we might end up with multiple backup URLs that are the same, isn't it?


At the same time, we need to be able `concat` to handle chunked uploads. 
When a chunked upload (>100MB) is in progress we hit POST `/car`  multiple times, which means we call `db.createUpload` multiple times and I'd assume the expected S3 files and therefore urls in the db should be
`/raw/ROOT_CID1/u1/CAR1.car, upload_key_1`
`/raw/ROOT_CID1/u1/CAR2.car, upload_key_1`


I guess we should, if possible, concat if  the url doesn't exists already. Which mimics in a way what we had before with 

```
ON CONFLICT ( upload_id, url ) DO NOTHING;
```


I've also noticed this is not in NFT.storage, we might have to port there whatever we figure out here, isn't it?

Reply:
That make sense. I made some changes to make backup URLs unique (a JSON doc with the URL as a key).